File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    
    bool isTransferable(int capacity, vector<int>& weights, int days) {
        
        // days need to transfer all items with limit of {capacity}
        int daysTaken = 1;
        
        // sum use for adding weights in one go
        int sum =0;
        
        for(int i =0; i< weights.size(); i++){
            
            sum += weights[i];
            
            // check sum is greater than the assumed capacity then daysTaken increase 
            // and new days start so sum start from current weight
            
            if(sum > capacity) {
                daysTaken ++;
                
                sum = weights[i];
            }
        }
        
        // if daysTaken is greater than provided days so it is not acceptable
        if(daysTaken <= days) {
            return true;
        }
        
        return false;
    }
    
    int shipWithinDays(vector<int>& weights, int days) {
        
        // max will be sum of all
        int maxCapacity=0;
        
        int n = weights.size();
        
        for(int i = 0; i < n; i++)
            maxCapacity += weights[i];
        
        int l = *max_element(weights.begin(), weights.end()), r = maxCapacity;
        int ans;
        
        while(l <= r) {
            
            // asumed capacity 
            int mid = l + (r - l) /2;
            
            // check in this capacity (mid) can we transfer total load in days
            // if yes then we go for least capacity
            
            if(isTransferable(mid, weights, days)){
                ans = mid;
                r = mid-1;    
            }else {
                l = mid +1 ;
            }
        }
        return ans;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cs
public class Solution {
/**
Instead of asking "What's the best partition?", flip it:
"If I give you a ship capacity of C, can you ship everything in D days?"
Think about this carefully. If the answer is YES for capacity C, what about capacity C+1?
Also YES! (If it fits in a smaller truck, it fits in a bigger one.)
If the answer is NO for capacity C, what about C-1?
Also NO! (If it doesn't fit in a big truck, it won't fit in a smaller one.)

*/

    public int ShipWithinDays(int[] weights, int days) {
        int left = 0, right = 0;
        
        // Set up search space
        foreach (int w in weights) {
            left = Math.Max(left, w);   // At least the heaviest package
            right += w;                 // At most the sum of all
        }
        
        // Binary search for minimum valid capacity
        while (left < right) {
            int mid = left + (right - left) / 2;
            
            if (CanShip(weights, days, mid)) {
                right = mid;      // mid works, try smaller
            } else {
                left = mid + 1;   // mid too small, need bigger
            }
        }
        
        return left;
    }
    
    private bool CanShip(int[] weights, int days, int capacity) {
        int requiredDays = 1;
        int currentLoad = 0;
        
        foreach (int w in weights) {
            if (currentLoad + w > capacity) {
                requiredDays++;
                currentLoad = w;
                if (requiredDays > days) return false;
            } else {
                currentLoad += w;
            }
        }
        
        return true;
    }
}